enhancements/microshift/microshift-apiserver-custom-certs.md

```diff
@@ -0,0 +1,191 @@
+---
+title: microshift-apiserver-custom-certs
+authors:
+  - "@eslutsky"
+reviewers:
+  - "@jerpeter, MicroShift contributor"
+  - "@pacevedom, MicroShift contributor"
+  - "@ggiguash, MicroShift contributor"
+  - "@benluddy, OpenShift API server team"
+  - "@standa, OpenShift Auth Team"
+
+approvers:
+  - "@jerpeter"
+api-approvers:
+  - N/A
+creation-date: 2021-01-18
+last-updated: 2023-01-18
+tracking-link:
+  - https://issues.redhat.com/browse/USHIFT-2101
+see-also:
+  - microshift-apiserver-certs.md
+replaces:
+  - N/A
+superseded-by:
+  - N/A
+---
+
+# MicroShift API server custom external certificates
+## Summary
+
+This enhancement extends the Microshift apiserver Certs to allow the user to
+configure the use of client certificates for api server external TLS access.
+
+
+> Anytime the document mentions API server, it refers to kube API server.
+
+
+## Motivation
+Currently, users of Microshift cannot provide their own generated certificates,
+Some customers have very strict requirements regarding TLS certs. There are frequently intermediate proxies which allow only TLS traffic with certs that are signed by organization owned CAs.
+
+### User Stories
+* As a MicroShift administrator, I want to be able to add  organization generated certificates.
+* As a MicroShift administrator, I want to provide additional DNS names for each certificate.
+* As a MicroShift administrator, I want to be able to remove organization generated certificates.
+* As a MicroShift administrator, I want to be able to remove DNS names for each certificate.
+
+### Goals
+* Allow MicroShift administrator to configure Microshift with externally generated certificates and
+domain names.
+
+### Non-Goals
+N/A
+
+## Proposal
+Proposal suggests to provide Administrator means adding their own Certificates using microshift configuration file.
+
+a new `apiServer` level section will be added to the configuration called `namedCertificates`:
+
+```yaml
+apiServer:
+  namedCertificates:
+  - certPath: ~/certs/api_fqdn_1.crt
+    keyPath:  ~/certs/api_fqdn_1.key
+  - certPath: ~/certs/api_fqdn_2.crt
+    keyPath:  ~/certs/api_fqdn_2.key
+    names:
+    - api_fqdn_1     
+    - *.apps.external.com
+
+```  
+
+for each provided certificate, The following configuration is proposed:
+1. `names` - optional list of explicit DNS names (leading wildcards allowed).
+   If no names are provided, the implicit names will be extracted from the certificates.
+1. `certPath` -  certificate full path.
+1. `keyPath` -   certificate key full path.
+
+certificate files will be read from their conifgured location by the Microshift service,
+each certification file will be validated (see validation rules).
+
+Those certificate will extend the default  API server [external](https://github.com/openshift/microshift/blob/main/pkg/controllers/kube-apiserver.go#L194) certificate configuration.
```

Comment:
Do we intend to allow self-signed certs to be provided? How about certs signed by public trust chains? Since we're not providing configuration for specifying the CA separately, are we requiring that the trust chain is appended to the certs that are provided by customers (e.g. so that we can populate the certificate-authority-data in a new kubeconfig?)

Reply:
well originally it was discussed removing certificate-authority-data from generated  custom kubeconfig altogether, 
so the Certs  will be validated against CAs that in the RHEL trust. customer can add certificate-authority-data manually if they want.

Reply:
added to proposal and the workflow  sections